src/main/java/com/commercetools/sync/services/impl/ChannelServiceImpl.java

```diff
@@ -0,0 +1,85 @@
+package com.commercetools.sync.services.impl;
+
+import com.commercetools.sync.services.ChannelService;
+import io.sphere.sdk.channels.Channel;
+import io.sphere.sdk.channels.ChannelDraft;
+import io.sphere.sdk.channels.ChannelDraftBuilder;
+import io.sphere.sdk.channels.ChannelRole;
+import io.sphere.sdk.channels.commands.ChannelCreateCommand;
+import io.sphere.sdk.channels.queries.ChannelQuery;
+import io.sphere.sdk.channels.queries.ChannelQueryBuilder;
+import io.sphere.sdk.client.SphereClient;
+import io.sphere.sdk.queries.QueryExecutionUtils;
+
+import javax.annotation.Nonnull;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+
+public class ChannelServiceImpl implements ChannelService {
+
+    private final SphereClient ctpClient;
+
+    /**
+     * Cache of Channels' [key -> id].
+     */
+    private final Map<String, String> cache = new HashMap<>();
+
+    public ChannelServiceImpl(@Nonnull final SphereClient ctpClient) {
+        this.ctpClient = ctpClient;
+    }
+
+    @Nonnull
+    @Override
+    public CompletionStage<Optional<String>> fetchCachedChannelIdByKeyAndRoles(@Nonnull final String key,
```

Comment:
very good catch! 68f4b44 I'd stick now to keeping a separate implementation of the service for each usage of the channel reference.